### <a name='0'></a>Import biblioteki

In [3]:
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

np.__version__

'1.24.4'

In [20]:
pd.__version__

'2.1.0'

### <a name='1'></a> Ćwiczenie 191
Z poniższego słownika _data_ utwórz obiekt _DataFrame_ i przypisz do zmiennej _df_.


```
data = {
    'size': ['XL', 'L', 'M', 'L', 'M'],
    'color': ['red', 'green', 'blue', 'green', 'red'],
    'gender': ['female', 'male', 'male', 'female', 'female'],
    'price': [199.0, 89.0, 99.0, 129.0, 79.0],
    'weight': [500, 450, 300, 380, 410],
    'bought': ['yes', 'no', 'yes', 'no', 'yes']
}
```



In [5]:
data = {
    'size': ['XL', 'L', 'M', 'L', 'M'],
    'color': ['red', 'green', 'blue', 'green', 'red'],
    'gender': ['female', 'male', 'male', 'female', 'female'],
    'price': [199.0, 89.0, 99.0, 129.0, 79.0],
    'weight': [500, 450, 300, 380, 410],
    'bought': ['yes', 'no', 'yes', 'no', 'yes']
}

df = pd.DataFrame(data)
df

size  color  gender  price  weight bought
0   XL    red  female  199.0     500    yes
1    L  green    male   89.0     450     no
2    M   blue    male   99.0     300    yes
3    L  green  female  129.0     380     no
4    M    red  female   79.0     410    yes

Wyświetl podstawowe informacje na temat obiektu _df_.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   size    5 non-null      object 
 1   color   5 non-null      object 
 2   gender  5 non-null      object 
 3   price   5 non-null      float64
 4   weight  5 non-null      int64  
 5   bought  5 non-null      object 
dtypes: float64(1), int64(1), object(4)
memory usage: 368.0+ bytes


Dokonaj konwersji zmiennych na poszczególne typy:
* size -> _category_
* color -> _category_
* gender -> _category_
* weight -> _float_
* bought -> _category_

In [4]:
for col in ['size', 'color', 'gender', 'bought']:
    df[col] = df[col].astype('category')

df['weight'] = df['weight'].astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   size    5 non-null      category
 1   color   5 non-null      category
 2   gender  5 non-null      category
 3   price   5 non-null      float64 
 4   weight  5 non-null      float64 
 5   bought  5 non-null      category
dtypes: category(4), float64(2)
memory usage: 740.0 bytes


### <a name='2'></a> Ćwiczenie 192
Wykorzystując bibliotekę do uczenia maszynowego _scikit-learn_ oraz klasę _LabelEncoder_ dokonaj kodowania 0-1 zmiennej _bought_. Zmiany przypisz na stałe do obiektu _df_ i wyświetl ten obiekt.

In [5]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['bought'] = le.fit_transform(df['bought'])
df

size  color  gender  price  weight  bought
0   XL    red  female  199.0   500.0       1
1    L  green    male   89.0   450.0       0
2    M   blue    male   99.0   300.0       1
3    L  green  female  129.0   380.0       0
4    M    red  female   79.0   410.0       1

Wykorzystując odpowiednią metodę klasy _LabelEncoder_ przywróć pierwotną postać zmiennej _bought_. Zwróć uwagę na typ zmiennej _bought_.

In [6]:
df['bought'] = le.inverse_transform(df['bought'])
df['bought'] = df['bought'].astype('category')
df

size  color  gender  price  weight bought
0   XL    red  female  199.0   500.0    yes
1    L  green    male   89.0   450.0     no
2    M   blue    male   99.0   300.0    yes
3    L  green  female  129.0   380.0     no
4    M    red  female   79.0   410.0    yes

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   size    5 non-null      category
 1   color   5 non-null      category
 2   gender  5 non-null      category
 3   price   5 non-null      float64 
 4   weight  5 non-null      float64 
 5   bought  5 non-null      category
dtypes: category(4), float64(2)
memory usage: 740.0 bytes


### <a name='3'></a> Ćwiczenie 193
Wykorzystując bibliotekę do uczenia maszynowego _scikit-learn_ oraz klasę _OneHotEncoder_ dokonaj kodowania 0-1 zmiennej _size_. Nie przypisuj zmian do obiektu _df_.

Użyj parametru _sparse=False_ klasy _OneHotEncoder_.

In [9]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(sparse=False)
encoder.fit_transform(df[['size']])

array([[0., 0., 1.],
       [1., 0., 0.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.]])

Wykorzystując bibliotekę do uczenia maszynowego _scikit-learn_ oraz klasę _OneHotEncoder_ dokonaj kodowania 0-1 zmiennej _size_. Nie przypisuj zmian do obiektu _df_.

Użyj parametru _sparse=False_ **oraz _drop=first_** klasy _OneHotEncoder_.

In [10]:
encoder = OneHotEncoder(sparse=False, drop='first')
encoder.fit_transform(df[['size']])

array([[0., 1.],
       [0., 0.],
       [1., 0.],
       [0., 0.],
       [1., 0.]])

### <a name='4'></a> Ćwiczenie 194
Wykorzystując bibliotekę _pandas_ oraz funkcję _get_dummies()_ przedstaw wszystkie zmienne kategoryczne obiektu _df_ w postaci kodowania 0-1. Nie przypisuj zmian na trwałe do obiektu _df_.

In [12]:
pd.get_dummies(df, dtype='int')

price  weight  size_L  size_M  size_XL  color_blue  color_green  color_red  \
0  199.0   500.0       0       0        1           0            0          1   
1   89.0   450.0       1       0        0           0            1          0   
2   99.0   300.0       0       1        0           1            0          0   
3  129.0   380.0       1       0        0           0            1          0   
4   79.0   410.0       0       1        0           0            0          1   

   gender_female  gender_male  bought_no  bought_yes  
0              1            0          0           1  
1              0            1          1           0  
2              0            1          0           1  
3              1            0          1           0  
4              1            0          0           1

Wykorzystując bibliotekę _pandas_ oraz funkcję _get_dummies()_ przedstaw wszystkie zmienne kategoryczne obiektu _df_ w postaci kodowania 0-1. Użyj parametru _drop_first=True_. Nie przypisuj zmian na trwałe do obiektu _df_.

In [13]:
pd.get_dummies(df, drop_first=True, dtype='int')

price  weight  size_M  size_XL  color_green  color_red  gender_male  \
0  199.0   500.0       0        1            0          1            0   
1   89.0   450.0       0        0            1          0            1   
2   99.0   300.0       1        0            0          0            1   
3  129.0   380.0       0        0            1          0            0   
4   79.0   410.0       1        0            0          1            0   

   bought_yes  
0           1  
1           0  
2           1  
3           0  
4           1

Wykorzystując bibliotekę _pandas_ oraz funkcję _get_dummies()_ przedstaw **zmienną _size_** obiektu _df_ w postaci kodowania 0-1. Użyj parametru _drop_first=True_. Nie przypisuj zmian na trwałe do obiektu _df_.

In [15]:
pd.get_dummies(df, columns=['size'], drop_first=True, dtype='int')

color  gender  price  weight bought  size_M  size_XL
0    red  female  199.0   500.0    yes       0        1
1  green    male   89.0   450.0     no       0        0
2   blue    male   99.0   300.0    yes       1        0
3  green  female  129.0   380.0     no       0        0
4    red  female   79.0   410.0    yes       1        0

### <a name='5'></a> Ćwiczenie 195
Wykorzystując bibliotekę do uczenia maszynowego _scikit-learn_ oraz klasę _StandardScaler_ dokonaj skalowania zmiennej _weight_. Nie przypisuj zmian do obiektu _df_.

In [18]:
#!pip install --upgrade scikit-learn

In [23]:
df

size  color  gender  price  weight bought
0   XL    red  female  199.0   500.0    yes
1    L  green    male   89.0   450.0     no
2    M   blue    male   99.0   300.0    yes
3    L  green  female  129.0   380.0     no
4    M    red  female   79.0   410.0    yes

In [22]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit_transform(df[['weight']])

array([[ 1.3660019 ],
       [ 0.62360956],
       [-1.60356745],
       [-0.41573971],
       [ 0.02969569]])

Standaryzacja jest ważnym krokiem w preprocessingu danych, szczególnie dla algorytmów uczących się maszynowo, które są czułe na skalę danych, takich jak regresja liniowa, SVM, czy k-najbliższych sąsiadów (k-NN). Standaryzacja zapewnia, że każda cecha ma tę samą skalę, co często prowadzi do lepszej wydajności modeli.

Standaryzacja cech (feature scaling) to technika przetwarzania danych stosowana w celu normalizacji wartości cech w zbiorze danych. Polega na przekształceniu cech tak, aby miały one podobną skalę, co jest szczególnie ważne w przypadku algorytmów uczenia maszynowego, które są wrażliwe na różnice w skali danych. Dwa najpopularniejsze rodzaje skalowania to standaryzacja i normalizacja.

### Standaryzacja

Standaryzacja cech polega na przekształceniu danych tak, aby miały średnią 0 i odchylenie standardowe 1. Wzór na standaryzację jest następujący:

\[ z = \frac{x - \mu}{\sigma} \]

gdzie:
- \( x \) jest wartością cechy,
- \( \mu \) jest średnią wartością cechy,
- \( \sigma \) jest odchyleniem standardowym cechy.

#### Przykład

Jeśli mamy zbiór danych z cechą `height` o średniej wartości 170 cm i odchyleniu standardowym 10 cm, to wartość 180 cm po standaryzacji wyniesie:

\[ z = \frac{180 - 170}{10} = 1 \]

### Normalizacja

Normalizacja cech polega na przekształceniu danych tak, aby wartości mieściły się w określonym przedziale, zwykle od 0 do 1. Wzór na normalizację (min-max scaling) jest następujący:

\[ x' = \frac{x - \min(x)}{\max(x) - \min(x)} \]

gdzie:
- \( x \) jest wartością cechy,
- \( \min(x) \) jest minimalną wartością cechy,
- \( \max(x) \) jest maksymalną wartością cechy.

#### Przykład

Jeśli mamy zbiór danych z cechą `height` z minimalną wartością 150 cm i maksymalną wartością 190 cm, to wartość 180 cm po normalizacji wyniesie:

\[ x' = \frac{180 - 150}{190 - 150} = \frac{30}{40} = 0.75 \]

### Dlaczego standaryzacja cech jest ważna?

1. **Ujednolicenie skali**: Wiele algorytmów uczenia maszynowego, takich jak k-najbliżsi sąsiedzi (k-NN), regresja liniowa, SVM, czy sieci neuronowe, działa lepiej, gdy cechy mają podobną skalę.
   
2. **Szybsza zbieżność**: Algorytmy optymalizacji, takie jak gradienty zstępujące, mogą szybciej osiągnąć zbieżność, gdy cechy są znormalizowane.

3. **Poprawa wydajności modeli**: Modele uczące się maszynowo mogą osiągać lepszą wydajność, gdy cechy są skalowane, ponieważ różnice w skali mogą prowadzić do niewłaściwych wag cech.

### Implementacja w Pythonie

Standaryzacja cech w Pythonie może być łatwo wykonana przy użyciu biblioteki `scikit-learn`:

```python
from sklearn.preprocessing import StandardScaler
import numpy as np

# Przykładowe dane
data = np.array([[1, 2, 3], 
                 [4, 5, 6], 
                 [7, 8, 9]])

# Tworzenie obiektu StandardScaler
scaler = StandardScaler()

# Dopasowanie i transformacja danych
scaled_data = scaler.fit_transform(data)

print("Przeskalowane dane:\n", scaled_data)
```

Wynik powyższego kodu to znormalizowane dane, gdzie każda cecha ma średnią 0 i odchylenie standardowe 1.

Podsumowując, standaryzacja cech jest kluczowym krokiem w przetwarzaniu danych, który pomaga w poprawie wydajności i dokładności modeli uczenia maszynowego.

W standaryzacji (standardization) cechy są przekształcane tak, aby miały średnią 0 i odchylenie standardowe 1. Maksymalne przedziały w standaryzacji nie są wyliczane w sposób bezpośredni, ponieważ standaryzacja skupia się na przekształceniu danych w taki sposób, że wartości przekształcone rozkładają się wokół średniej 0 z jednostkowym odchyleniem standardowym.

Jednak aby zrozumieć, jakie wartości może przyjmować cecha po standaryzacji, możemy przyjrzeć się wartościom skrajnym w oryginalnym zbiorze danych oraz sposobowi ich przekształcenia:

### Kroki standaryzacji

1. **Obliczenie średniej (\(\mu\))**: Średnia wszystkich wartości danej cechy.
2. **Obliczenie odchylenia standardowego (\(\sigma\))**: Miara rozproszenia wartości cechy wokół średniej.
3. **Przekształcenie wartości (\(z\))**: Każda wartość \(x\) jest przekształcana przy użyciu wzoru:

\[ z = \frac{x - \mu}{\sigma} \]

### Interpretacja przekształconych wartości

Po standaryzacji wartości przekształcone (\(z\)) będą rozkładać się wokół 0 z jednostkowym odchyleniem standardowym. Wartości teoretycznie mogą przyjmować dowolne wartości w zależności od rozkładu oryginalnych danych, ale w praktyce dla większości danych rozkład normalny obejmuje wartości w przedziale \([-3, 3]\) około 99.7% przypadków (zgodnie z regułą trzech sigm).

### Przykład

Rozważmy prosty przykład:

Oryginalne dane:
\[ \text{data} = [1, 2, 3, 4, 5] \]

1. **Obliczenie średniej (\(\mu\))**:
\[ \mu = \frac{1 + 2 + 3 + 4 + 5}{5} = 3 \]

2. **Obliczenie odchylenia standardowego (\(\sigma\))**:
\[ \sigma = \sqrt{\frac{(1-3)^2 + (2-3)^2 + (3-3)^2 + (4-3)^2 + (5-3)^2}{5}} = \sqrt{\frac{4 + 1 + 0 + 1 + 4}{5}} = \sqrt{2} \approx 1.414 \]

3. **Standaryzacja wartości**:
\[ z = \frac{x - \mu}{\sigma} \]

Przekształcone wartości:
\[ z = \left[ \frac{1-3}{1.414}, \frac{2-3}{1.414}, \frac{3-3}{1.414}, \frac{4-3}{1.414}, \frac{5-3}{1.414} \right] = [-1.414, -0.707, 0, 0.707, 1.414] \]

### Maksymalne przedziały

W przypadku powyższego przykładu, maksymalne wartości po standaryzacji to około \(\pm1.414\). W rzeczywistości, maksymalne wartości po standaryzacji będą zależeć od rozkładu danych i mogą przyjmować szerszy zakres dla bardziej zróżnicowanych danych.

### Podsumowanie

- **Standaryzacja** przekształca cechy tak, aby miały średnią 0 i odchylenie standardowe 1.
- **Przekształcone wartości** mogą teoretycznie przyjmować szeroki zakres wartości, ale w praktyce dla większości danych będą mieścić się w przedziale \([-3, 3]\).
- **Maksymalne przedziały** po standaryzacji są zależne od rozkładu oryginalnych danych i nie są wyliczane bezpośrednio, ale wynikają z przekształcenia danych względem ich średniej i odchylenia standardowego.

### <a name='6'></a> Ćwiczenie 196
Wczytaj raz jeszcze słownik _data_ do obiektu _DataFrame_ o nazwie _df_.

In [6]:
df = pd.DataFrame(data)
df

size  color  gender  price  weight bought
0   XL    red  female  199.0     500    yes
1    L  green    male   89.0     450     no
2    M   blue    male   99.0     300    yes
3    L  green  female  129.0     380     no
4    M    red  female   79.0     410    yes

Przygotuj dane do modelu uczenia maszynowego:
* wykorzystując klasę LabelEncoder dokonaj kodowania 0-1 zmiennej _bought_
* dokonaj standaryzacji zmiennych _weight_ oraz _price_
* dokonaj kodowania 0-1 posoztałych zmiennych kategorycznych wykorzystując funkcję _get_dummies()._

Przypisz zmiany do zmiennej _df_prep_.

In [8]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

In [9]:
df_prep = df.copy()

encoder = LabelEncoder()
df_prep['bought'] = encoder.fit_transform(df_prep['bought'])

scaler = StandardScaler()
df_prep[['price', 'weight']] = scaler.fit_transform(df_prep[['price', 'weight']])

df_prep = pd.get_dummies(df_prep, drop_first=True, dtype=int)
df_prep

price    weight  bought  size_M  size_XL  color_green  color_red  \
0  1.845062  1.366002       1       0        1            0          1   
1 -0.691898  0.623610       0       0        0            1          0   
2 -0.461266 -1.603567       1       1        0            0          0   
3  0.230633 -0.415740       0       0        0            1          0   
4 -0.922531  0.029696       1       1        0            0          1   

   gender_male  
0            0  
1            1  
2            1  
3            0  
4            0

### <a name='7'></a> Ćwiczenie 197
Skopiuj obiekt _df_prep_ do zmiennej _data_. Następnie wyrwij z tego obiektu kolumnę _bought_ i przypisz do zmiennej _target_.


In [10]:
data = df_prep.copy()
target = data.pop('bought')

Wyświetl 5 pierwszych wierszy obiektu _data_. Upewnij się, że nie ma w nim kolumny _bought_.

In [11]:
data.head()

price    weight  size_M  size_XL  color_green  color_red  gender_male
0  1.845062  1.366002       0        1            0          1            0
1 -0.691898  0.623610       0        0            1          0            1
2 -0.461266 -1.603567       1        0            0          0            1
3  0.230633 -0.415740       0        0            1          0            0
4 -0.922531  0.029696       1        0            0          1            0

Wyświetl 5 pierwszych wierszy obiektu _target_.

In [12]:
target.head()

0    1
1    0
2    1
3    0
4    1
Name: bought, dtype: int32

### <a name='8'></a> Ćwiczenie 198
Wykorzystując bibliotekę _scikit-learn_ oraz funkcję _train_test_split_ podziel dane z obiektów _data_ i _target_ na zbiór treningowy (_X\_train_, _y\_train_) oraz zbiór testowy (_X\_test_, _y\_test_). Do zbioru testowego przydziel 20% próbek.

Wyświetl rozmiar obiektów:
* _X\_train_
* _X\_test_
* _y\_train_
* _y\_test_

In [14]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2)

print(f'X_trian shape: {X_train.shape}')
print(f'X_test shape: {X_test.shape}')
print(f'y_trian shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')

X_trian shape: (4, 7)
X_test shape: (1, 7)
y_trian shape: (4,)
y_test shape: (1,)


### <a name='9'></a> Ćwiczenie 199
Połącz obiekty _X\_train_ oraz _X\_test_ w jeden obiekt _DataFrame_ o nazwie _X_.

In [18]:
X = pd.concat([X_train, X_test])
X

price    weight  size_M  size_XL  color_green  color_red  gender_male
0  1.845062  1.366002       0        1            0          1            0
3  0.230633 -0.415740       0        0            1          0            0
1 -0.691898  0.623610       0        0            1          0            1
2 -0.461266 -1.603567       1        0            0          0            1
4 -0.922531  0.029696       1        0            0          1            0

### <a name='10'></a> Ćwiczenie 200
Połącz obiekty _y\_train_ oraz _y\_test_ w jeden obiekt _DataFrame_ o nazwie _y_.

In [19]:
y = pd.concat([y_train, y_test])
y

0    1
3    0
1    0
2    1
4    1
Name: bought, dtype: int32

Połącz obiekty _X_ oraz _y_ w jeden obiekt _DataFrame_.

In [20]:
pd.concat([X, y], axis=1)

price    weight  size_M  size_XL  color_green  color_red  gender_male  \
0  1.845062  1.366002       0        1            0          1            0   
3  0.230633 -0.415740       0        0            1          0            0   
1 -0.691898  0.623610       0        0            1          0            1   
2 -0.461266 -1.603567       1        0            0          0            1   
4 -0.922531  0.029696       1        0            0          1            0   

   bought  
0       1  
3       0  
1       0  
2       1  
4       1